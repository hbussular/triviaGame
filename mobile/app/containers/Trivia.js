import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View } from 'react-native'

import PrimaryButton from '../components/PrimaryButton'
import QuestionHeader from '../components/QuestionHeader'
import QuestionCard from '../components/QuestionCard'
import QuestionOption from '../components/QuestionOption'

import { calculateScore, prepareQuestion } from '../engines/triviaEngine'

import styles from './Trivia.style'

class Trivia extends Component {
  constructor(props) {
    super(props)

    this.state = {
      questionNumber: 0,
      questions: [],
      question: {},
      score: 0,
    }
  }

  componentWillMount() {
    const {
      questions,
    } = this.props

    const {
      questionNumber,
    } = this.state

    this.setState({
      questions,
      question: this.getQuestion(questions, questionNumber),
    })
  }

  getQuestion = (questions, questionNumber) => prepareQuestion(questions[questionNumber])

  nextQuestion = () => {
    const {
      questions,
      questionNumber,
      score,
    } = this.state

    const {
      onFinish,
    } = this.props

    if (questionNumber < questions.length - 1) {
      const nextQuestionNumber = questionNumber + 1
      const newQuestion = this.getQuestion(questions, nextQuestionNumber)

      this.setState({
        questionNumber: nextQuestionNumber,
        question: newQuestion,
      })
    } else {
      onFinish({ score })
    }
  }

  handlePress = ({ correctAwnser }) => {
    const {
      questions,
      questionNumber,
      score,
    } = this.state

    if (correctAwnser) {
      const question = questions[questionNumber]
      const questionDifficulty = question.difficulty

      const finalScore = calculateScore(score, questionDifficulty)

      this.setState({
        score: finalScore,
      })
    }

    this.nextQuestion()
  }

  render() {
    const {
      onLeave,
    } = this.props

    const {
      questionNumber,
      questions,
      question,
    } = this.state

    return (
      <View style={styles.triviaContainer}>
        <View style={styles.headerContainer}>
          <QuestionHeader
            questionPosition={questionNumber + 1}
            questionsLength={questions.length}
          />
        </View>
        <View style={styles.questionContainer}>
          <QuestionCard question={question.question} />
        </View>
        <View style={styles.optionsContainer}>
          {
            question.options.map((item, idx) => (
              <QuestionOption
                answer={item.answer}
                correct={item.correct}
                key={idx} // eslint-disable-line react/no-array-index-key
                onPress={this.handlePress}
              />
            ))
          }
        </View>
        <View style={styles.bottomContainer}>
          <PrimaryButton
            label="Leave"
            onPress={onLeave}
          />
        </View>
      </View>
    )
  }
}

Trivia.propTypes = {
  onFinish: PropTypes.func.isRequired,
  onLeave: PropTypes.func.isRequired,
  questions: PropTypes.array.isRequired,
}

export default Trivia
