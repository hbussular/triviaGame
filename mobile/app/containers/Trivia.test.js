import React from 'react'
import renderer from 'react-test-renderer'
import Trivia from './Trivia'

import { triviaScenario } from '../tests/scenarios'

test('renders correctly', () => {
  const tree = renderer
    .create(<Trivia {...triviaScenario} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})
