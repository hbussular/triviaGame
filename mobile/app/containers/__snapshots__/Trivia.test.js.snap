// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "height": "100%",
    }
  }
>
  <View
    style={Object {}}
  >
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "rgba(255, 255, 255, 1)",
            "fontSize": 38,
            "fontWeight": "700",
            "padding": 20,
            "textAlign": "center",
          }
        }
      >
        Question 1 of 1
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "height": "30%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 8,
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "padding": 30,
          "width": "100%",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        Which "Toy Story" character was voiced by Don Rickles?
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "height": "40%",
        "paddingTop": 20,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderRadius": 8,
          "marginBottom": 5,
          "opacity": 1,
          "paddingVertical": 20,
          "width": "100%",
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "rgba(213, 55, 43, 1)",
              "fontSize": 16,
              "fontWeight": "700",
              "textAlign": "center",
            },
            Object {
              "color": "rgba(213, 55, 43, 1)",
            },
          ]
        }
      >
        Mr. Potato Head
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderRadius": 8,
          "marginBottom": 5,
          "opacity": 1,
          "paddingVertical": 20,
          "width": "100%",
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "rgba(213, 55, 43, 1)",
              "fontSize": 16,
              "fontWeight": "700",
              "textAlign": "center",
            },
            Object {
              "color": "rgba(213, 55, 43, 1)",
            },
          ]
        }
      >
        Slinky Dog
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderRadius": 8,
          "marginBottom": 5,
          "opacity": 1,
          "paddingVertical": 20,
          "width": "100%",
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "rgba(213, 55, 43, 1)",
              "fontSize": 16,
              "fontWeight": "700",
              "textAlign": "center",
            },
            Object {
              "color": "rgba(213, 55, 43, 1)",
            },
          ]
        }
      >
        Hamm
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderRadius": 8,
          "marginBottom": 5,
          "opacity": 1,
          "paddingVertical": 20,
          "width": "100%",
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "rgba(213, 55, 43, 1)",
              "fontSize": 16,
              "fontWeight": "700",
              "textAlign": "center",
            },
            Object {
              "color": "rgba(213, 55, 43, 1)",
            },
          ]
        }
      >
        Rex
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "bottom": 0,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "position": "absolute",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderRadius": 8,
          "display": "flex",
          "opacity": 1,
          "paddingVertical": 20,
          "width": "100%",
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontWeight": "700",
              "textAlign": "center",
            },
            Object {
              "color": "rgba(213, 55, 43, 1)",
            },
            Object {
              "fontSize": 16,
            },
          ]
        }
      >
        Leave
      </Text>
    </View>
  </View>
</View>
`;
